core/src/cgroup.rs

```diff
@@ -1,238 +1,392 @@
+        // We need to check for the existance of cgroup.kill, because
+        // this file does not exist on the root cgroup.
+        let path = self.path.join("cgroup.kill");
+        if !path.exists() {
+            bail!("cannot kill the root cgroup");
         }
+
+        Ok(fs::write(path, "1")?)
     }
 
-    /// Deletes a cgroup by killing it's processes and removing the directory afterwards
-    pub fn delete(&self, timeout: Duration) -> TypedResult<()> {
-        self.kill_all_timeout(timeout)?;
-        for d in WalkDir::new(&self.path)
-            .into_iter()
-            .flatten()
-            .filter(|e| e.file_type().is_dir())
-            .sorted_by(|a, b| a.depth().cmp(&b.depth()).reverse())
-        {
-            std::fs::remove_dir(&d.path()).typ(SystemError::CGroup)?;
-            trace!("Removed {:?}", d.path().as_os_str())
+    /// Returns the path of this cgroup
+    pub fn get_path(&self) -> PathBuf {
+        self.path.clone()
+    }
+
+    /// Returns the path of the event file, which may be polled
+    pub fn get_events_path(&self) -> PathBuf {
+        self.path.join("cgroup.events")
+    }
+
+    /// Kills all processes and removes the current cgroup
+    pub fn rm(&self) -> anyhow::Result<()> {
+        if !is_cgroup(&self.path)? {
+            bail!(
+                "{} is not a valid cgroup",
+                self.path.to_str().unwrap_or("N/A")
+            );
+        }
+
+        // Recursively delete all sub cgroups
+        for entry in fs::read_dir(&self.path)? {
+            let entry = entry?;
+            let meta = entry.metadata()?;
+
+            if meta.is_dir() {
+                // Each sub cgroup is deleted by calling this function recursively
+                let child = Self::import_root(&entry.path())?;
+                child.rm()?;
+            }
         }
+
+        // Remove the actual cgroup
+        self.kill()?;
+        while self.populated()? {}
+        fs::remove_dir(&self.path)?;
+
         Ok(())
     }
 
-    /// Returns the absolute path of this cgroup on the host file system
-    pub fn path(&self) -> PathBuf {
-        self.path.clone()
+    // TODO: Implement functions to fetch the parents and children
+}
+
+/// Returns the first cgroup2 mount point found on the host system
+pub fn mount_point() -> anyhow::Result<PathBuf> {
+    // TODO: This is an awful old function, replace it!
+    procfs::process::Process::myself()?
+        .mountinfo()?
+        .iter()
+        .find(|m| m.fs_type.eq("cgroup2")) // TODO A process can have several cgroup mounts
+        .ok_or_else(|| anyhow!("no cgroup2 mount found"))
+        .map(|m| m.mount_point.clone())
+}
+
+/// Checks if path is a valid cgroup by comparing the device id
+fn is_cgroup(path: &Path) -> anyhow::Result<bool> {
+    let st = statfs::statfs(path)?;
+    Ok(st.filesystem_type() == statfs::CGROUP2_SUPER_MAGIC)
+}
+
+#[cfg(test)]
+mod tests {
+    // The tests must be run as root with --test-threads=1
```

Comment:
I tried using `systemd-run --user --scope` instead and it doesn't work, because I overlooked that you create test cgroups directly on-top of the cgroup mount-point which is always owned by root.
However, `systemd-run --user --scope` runs all tests successfully if all tests create cgroups in the cgroup of the running process.  
If we opt for this, we should really consider adding the current-process-cgroup-path-find(:tm:) function to our `CGroup` struct

Reply:
>  If we opt for this, we should really consider adding the current-process-cgroup-path-find(tm) function to our CGroup struct

The latter one makes sense to me.

Reply:
Done. See 831ea4cf71c3329d1d0697253caa1c8ead1b38be